### DISCUSSION TITLE
Synchronous widget builds or other way to avoid missing data during initial widget build?

### DISCUSSION THREAD

**Discussion by willsmanley** [simolus3/drift#3376] [General]
I have a widget that depends on a row of data in a drift table.

So I'm using a StreamBuilder.

However the first build always has the data missing, even though it is present in the table.

I specifically don't want to have any loading state for this widget.

Any way to force synchronicity so that widget builds always get the data if it exists? Or any other strategy?

**Comment by simolus3** [comment]
Since queries in drift are fundamentally running asynchronously (which is generally a good thing, because it allows us to run them on a background isolate without blocking the UI), there will always be a short loading state.

If it's crucial that there is no loading even for a single frame (e.g. because of a `Hero` transition), a possible solution might be to use `initialData` on the `StreamBuilder`. That data would be initialized by running the query once (as a future with `.get()`) and then passing the results to the pushed page.

**Reply by willsmanley** [reply]
That makes sense, thank you.

I'm trying to imagine an abstraction in which introducing a new widget always does the initial query, uses it as initialData, then subscribes to updates.

Reason being is this is a very common use pattern for me in basically every widget, especially for a local-first app. There aren't many loading states because the client has all of the data most of the time. And I have a liberal use of heroes.

I'll update here if I come up with something good.

**Reply by AhsanSarwar45** [reply]
> That makes sense, thank you.
> 
> I'm trying to imagine an abstraction in which introducing a new widget always does the initial query, uses it as initialData, then subscribes to updates.
> 
> Reason being is this is a very common use pattern for me in basically every widget, especially for a local-first app. There aren't many loading states because the client has all of the data most of the time. And I have a liberal use of heroes.
> 
> I'll update here if I come up with something good.

Hey I'm in a similar situation, which method best suited your case?
